tests/nvmeof/test_ceph_nvmeof_neg_tests.py

```diff
@@ -5,154 +5,398 @@
-import time
-from concurrent.futures import ThreadPoolExecutor
 from copy import deepcopy
 from time import sleep
 
 from ceph.ceph import Ceph, SocketTimeoutException
 from ceph.ceph_admin import CephAdmin
 from ceph.ceph_admin.helper import check_service_exists
-from ceph.ceph_admin.orch import Orch
 from ceph.parallel import parallel
 from ceph.rados.core_workflows import RadosOrchestrator
 from ceph.rados.monitor_workflows import MonitorWorkflows
 from ceph.rbd.workflows.cluster_operations import operation, osd_remove_and_add_back
-from ceph.utils import get_node_by_id, get_nodes_by_ids
-from cli.cephadm.cephadm import CephAdm
-from cli.utilities.utils import get_running_containers, reboot_node
-from tests.cephadm import test_nvmeof, test_orch
-from tests.nvmeof.test_ceph_nvmeof_gateway import (
+from ceph.utils import get_node_by_id
+from cli.utilities.utils import reboot_node
+from tests.cephadm import test_orch
+from tests.nvmeof.test_ceph_nvmeof_gateway import initiators
+from tests.nvmeof.workflows.gateway_entities import (
+    configure_hosts,
+    configure_listeners,
     configure_subsystems,
-    initiators,
     teardown,
 )
 from tests.nvmeof.workflows.initiator import NVMeInitiator
-from tests.nvmeof.workflows.nvme_gateway import create_gateway
-from tests.nvmeof.workflows.nvme_utils import (
-    check_and_set_nvme_cli_image,
-    deploy_nvme_service,
-    nvme_gw_cli_version_adapter,
-    setup_firewalld,
-)
+from tests.nvmeof.workflows.nvme_service import NVMeService
+from tests.nvmeof.workflows.nvme_utils import check_and_set_nvme_cli_image
 from tests.rbd.rbd_utils import initial_rbd_config
 from utility.log import Log
 from utility.retry import retry
-from utility.utils import find_free_port, generate_unique_id
+from utility.utils import generate_unique_id
 
 LOG = Log(__name__)
 cli_image = str()
 
 
-def test_ceph_83575812(ceph_cluster, rbd, pool, config):
+def test_ceph_83575812(ceph_cluster, rbd, nvme_service, pool, config):
     """CEPH-83575812 Remove the image during NVMe images are in use."""
-    gw_node = get_node_by_id(ceph_cluster, config["gw_node"])
-    ceph = Orch(ceph_cluster, **{})
-    nvmegwcli = create_gateway(
-        nvme_gw_cli_version_adapter(ceph_cluster),
-        gw_node,
-        mtls=config.get("mtls"),
-        shell=getattr(ceph, "shell"),
-        port=config.get("gw_port", 5500),
-        gw_group=config.get("gw_group"),
-    )
-    listener_port = find_free_port(gw_node)
-    subsystem = dict()
-    subsystem.update(
-        {
-            "nqn": "nqn.2016-06.io.spdk:negative",
-            "serial": 114,
-            "listener_port": listener_port,
-            "allow_host": "*",
+    try:
+        # Deploy NVMe Service
+        if config.get("install"):
+            nvme_service.deploy()
+
+        # Initialize gateways
+        nvme_service.init_gateways()
+        nvmegwcli = nvme_service.gateways[0]
+
+        # Configure Subsystems
+        configure_subsystems(nvme_service)
```

Comment:
Can we not directly call configure_gw_entities and that will take care of configuring subsystems, listeners and hosts right? All you need to do is pass those keywords to config I guess. Can we check this?